hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java

```diff
@@ -159,41 +168,73 @@ public static synchronized void executeJar(@Nonnull Supplier<HazelcastInstance>
 
             Class<?> clazz = loadMainClass(classLoader, mainClass);
 
-            Method main = clazz.getDeclaredMethod("main", String[].class);
-            int mods = main.getModifiers();
-            if ((mods & Modifier.PUBLIC) == 0 || (mods & Modifier.STATIC) == 0) {
-                error("Class " + clazz.getName()
-                        + " has a main(String[] args) method which is not public static");
-            }
+            Method main = getMainMethod(clazz, calledByMember);
             String[] jobArgs = args.toArray(new String[0]);
             // upcast args to Object so it's passed as a single array-typed argument
             main.invoke(null, (Object) jobArgs);
-            awaitJobsStarted();
+
+            // Wait for the job to start only if called by the client side
+            if (!calledByMember) {
+                awaitJobsStarted();
+            }
+
         } finally {
-            HazelcastInstance remembered = HazelcastBootstrap.supplier.remembered();
-            if (remembered != null) {
-                try {
-                    remembered.shutdown();
-                } catch (Throwable t) {
-                    System.err.println("Shutdown failed with:");
-                    t.printStackTrace();
+            // HazelcastInstance should be closed if called by client side
+            // HazelcastInstance should not be closed if called by member side
+            if (!calledByMember) {
+                HazelcastInstance remembered = HazelcastBootstrap.supplier.remembered();
+                if (remembered != null) {
+                    try {
+                        remembered.shutdown();
+                    } catch (Throwable t) {
+                        System.err.println("Shutdown failed with:");
+                        t.printStackTrace();
+                    }
                 }
+                resetSupplier();
             }
-            HazelcastBootstrap.supplier = null;
+        }
+    }
+
+    private static Method getMainMethod(Class<?> clazz, boolean calledByMember) throws NoSuchMethodException {
+        // If main method does not exist, throws NoSuchMethodException
+        Method main = clazz.getDeclaredMethod("main", String[].class);
+
+        int mods = main.getModifiers();
+        if ((mods & Modifier.PUBLIC) == 0 || (mods & Modifier.STATIC) == 0) {
```

Comment:
```suggestion
        if (Modifier.isPublic(mods) || Modifier.isStatic(mods)) {
```

For even better clarity you can then import static isPublic and isStatic

Reply:
Ok